bpf/tc/attach.go

```diff
@@ -105,6 +105,19 @@ func (ap AttachPoint) AttachProgram() error {
 		return err
 	}
 
+	matchedHash, objHash, err := bpf.CheckAttachedProgs(ap.IfaceName(), ap.FileName())
+	if err == nil {
+		for _, p := range progsToClean {
```

Comment:
In the mainline case there should be exactly one program in `progsToClean` if the previous felix got as far as `RememberAttachedProgs`.  So, rather than checking `p.object`, I suggest that you check `len(progsToClean) == 1`.  If the length is 1, trust that the program is correct; otherwise, fall through and go through the full clean-up/reapply logic.

I think that avoids risks around `p.object` taking an unexpected value.  Why do I think that's a risk?  ISTR that the value that you'll see in p.object has changed across kernels and libbpf might do it differently.  (Ultimately, it's just a string that userspace tells the kernel to print.)